### PR TITLE
BigQuery Storage Write API SchemaTransform wrapper for Python SDK

### LINKED ISSUE #21961: [Feature Request]: Add support for BigQuery Storge Write API in Python
### What would you like to happen?

Much like #20687 to read to BigQuery, The Java SDK contains an implementation which uses the BigQuery Storage API to write to BigQuery. We should implement the same for Python.

In our case, we would like to use it instead of using the `FILE_LOADS` ingestion method in streaming mode, where errors during the loading process are not propagated to the pipeline. We are currently using it as we do not need `STREAMING_INSERT` features, and hence want to save on cost.

### Issue Priority

Priority: 3

### Issue Component

Component: io-py-gcp

### PR DESCRIPTION
**Pull request by ahmedabu98**
Implementing a wrapper for Python SDK that uses the Storage API SchemaTransform (#23988) to write to BigQuery

Fixes #21961

### CODE REVIEW

**Comment by ahmedabu98** [comment]
R: @chamikaramj

**Review by chamikaramj — COMMENTED** [review]
Thanks. Looks great. Just nits.

**Comment by chamikaramj** [comment]
Did you try triggering the Jenkins integration test suites for the test with this PR ?

File: sdks/python/apache_beam/transforms/external.py [hunk 1]
```diff
@@ -367,6 +367,32 @@ def discover(expansion_service):
             inputs=proto_config.input_pcollection_names,
             outputs=proto_config.output_pcollection_names)
 
+  @staticmethod
+  def discover_one(expansion_service, name):
```
[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/python/apache_beam/transforms/external.py:371]: sure SGTM

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java [hunk 2]
```diff
@@ -691,6 +692,12 @@ public static Row toBeamRow(Schema rowSchema, TableSchema bqSchema, TableRow jso
       if (JSON_VALUE_PARSERS.containsKey(fieldType.getTypeName())) {
         return JSON_VALUE_PARSERS.get(fieldType.getTypeName()).apply(jsonBQString);
       } else if (fieldType.isLogicalType(SqlTypes.DATETIME.getIdentifier())) {
+        // Handle if datetime value is in micros
```
[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java:695]: Thanks for the explanation.

Sorry, by log, I meant a comment :)

File: sdks/python/apache_beam/transforms/external.py [hunk 3]
```diff
@@ -367,6 +367,32 @@ def discover(expansion_service):
...
+    """Discover one SchemaTransform by name in the given expansion service.
+
+    :return: one SchemaTransformConfig that represents the discovered
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/external.py:374]: Also, mention what the function would do if the transform is not found.

**Comment by ahmedabu98** [comment]
@chamikaramj let me know if the mock service unit test is along the lines of what you were thinking

File: sdks/python/apache_beam/io/gcp/bigquery_write_it_test.py [hunk 4]
```diff
@@ -545,6 +548,102 @@ def test_big_query_write_temp_table_append_schema_update(self, file_format):
...
+    self.project = self.test_pipeline.get_option('project')
+
+    self.bigquery_client = BigQueryWrapper()
+    self.dataset_id = '%s%s%d' % (
+        self.BIGQUERY_DATASET, str(int(time.time())), random.randint(0, 10000))
+    self.bigquery_client.get_or_create_dataset(self.project, self.dataset_id)
+    _LOGGER.info(
+        "Created dataset %s in project %s", self.dataset_id, self.project)
+
+    self.expansion_service = ('localhost:%s' % os.environ.get('EXPANSION_PORT'))
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery_write_it_test.py:597]: We should be using this expansion service when running the pipeline ?

Also, does this service already get created ?

**Review by chamikaramj — COMMENTED** [review]
Thanks. I meant to try test_xlang_storage_write using a Jenkins trigger.

**Comment by ahmedabu98** [comment]
Run Python 3.8 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.7 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.10 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.7 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.10 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.7 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.7 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.10 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.7 PostCommit

**Comment by ahmedabu98** [comment]
Run XVR_PythonUsingJava_Dataflow PostCommit

**Comment by ahmedabu98** [comment]
Closing this PR in favor of #25521 as this one has gotten too messy.

**Pull request closed without merging**